gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -0,0 +1,300 @@
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.Job;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import gobblin.compaction.Compactor;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+
+
+/**
+ * MapReduce-based {@link gobblin.compaction.Compactor}.
+ *
+ * @author ziliu
+ */
+public class MRCompactor implements Compactor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MRCompactor.class);
+  private static final List<Job> RUNNING_MR_JOBS = Lists.newCopyOnWriteArrayList();
+
+  private final State state;
+  private final String inputDir;
+  private final String inputSubDir;
+  private final String destDir;
+  private final String destSubDir;
+  private final String tmpDir;
+  private final FileSystem fs;
+  private final ExecutorService executorService;
+  private final List<Future<?>> futures;
+
+  public MRCompactor(Properties props) throws IOException {
+    this.state = new State();
+    state.addAll(props);
+    this.inputDir = getInputDir();
+    this.inputSubDir = getInputSubDir();
+    this.destDir = getDestDir();
+    this.destSubDir = getDestSubDir();
+    this.tmpDir = getTmpDir();
+    this.fs = getFileSystem();
+    this.executorService = createExecutorService();
+    this.futures = Lists.newArrayList();
+  }
+
+  private String getInputDir() {
+    Preconditions.checkArgument(this.state.contains(ConfigurationKeys.COMPACTION_INPUT_DIR), "Property "
+        + ConfigurationKeys.COMPACTION_INPUT_DIR + " not provided");
+    return this.state.getProp(ConfigurationKeys.COMPACTION_INPUT_DIR);
+  }
+
+  private String getInputSubDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_INPUT_SUBDIR,
+        ConfigurationKeys.DEFAULT_COMPACTION_INPUT_SUBDIR);
+  }
+
+  private String getDestDir() {
+    Preconditions.checkArgument(this.state.contains(ConfigurationKeys.COMPACTION_DEST_DIR), "Property "
+        + ConfigurationKeys.COMPACTION_DEST_DIR + " not provided");
+    return this.state.getProp(ConfigurationKeys.COMPACTION_DEST_DIR);
+  }
+
+  private String getDestSubDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_DEST_SUBDIR,
+        ConfigurationKeys.DEFAULT_COMPACTION_DEST_SUBDIR);
+  }
+
+  private String getTmpDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_TMP_DIR, ConfigurationKeys.DEFAULT_COMPACTION_TMP_DIR);
+  }
+
+  private FileSystem getFileSystem() throws IOException {
+    Configuration conf = new Configuration();
+    for (String propName : this.state.getPropertyNames()) {
+      conf.set(propName, this.state.getProp(propName));
+    }
+    return FileSystem.get(conf);
```

Comment:
We can use gobblin configuration key "fs.uri" or other more specific keys, e.g. "writer.fs.uri" to get our uri here.